core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerProxy.scala

```diff
@@ -870,6 +962,71 @@ object ContainerProxy {
   }
 }
 
+object TCPPingClient {
+  def props(containerId:String, config: ContainerProxyHealthCheckConfig, remote: InetSocketAddress, replies: ActorRef) =
+    Props(new TCPPingClient(containerId, remote, replies, config))
+}
+
+class TCPPingClient(containerId:String, remote: InetSocketAddress, listener: ActorRef, config: ContainerProxyHealthCheckConfig)
```

Comment:
If this `PingActor` is only used to keepalive prewarm containers, how about making it be shared among containers, rather than having dedicated ping actors for each container?
Then this becomes an invoker-wide functionality to keep prewarm containers healthy.
And I think it would not be a big overhead for this actor to send a ping to multiple containers.

Reply:
I don't think this will work well. The ContainerProxy actor is already a 1:1 actor per container, and the state in ContainerProxy dictates when it should and should not be active. The overhead of an actor is minimal.

Reply:
Yes, I meant an invoker-wide service that keeps prewarm containers healthy.
So we can have 1:N mapping between PingingService and prewarm containers while still keeping the 1:1 mapping between ContainerProxy and a real container.

With this, PingingService doesn't need to be stopped and created upon the creation/deletion of ContainerProxy.

Each a prewarm ContainerProxy can request this service to enable the ping and PingingService can have a map or list of target endpoints to send a ping.

Reply:
I don't think having a separate service will be simpler, since a) it will need to track the internal state of ContainerProxy and b) that state is already tracked inside ContainerProxy and c) we would still create and destroy TCPPingClient and IO(Tcp) actors, which should also be tied to the ContainerProxy state (should be deleted when ContainerProxy is deleted etc)

Reply:
What I suggested is,

0) the service will be created on the bootstrap of invokers and the ActorRef will be sent to each (prewarm) container.
1) Prewarm containers will send a registration request to the service.
2) When it becomes other than the prewarm state, then it will send a deregistration request to the service.
3) The service will run forever as long as a corresponding invoker is running.
4) Once the service receives the registration request, it can store the endpoint in a map or list.
5) Every interval, it will send ping requests to all endpoints in the map.

So a) it doesn't need to track the internal state of ContainerProxy, b) only prewarm containers will register themselves on the creation and deregister when being warmed. so now the service is independent of the ContainerProxy state. c) it does not need to be removed as there will be always some prewarm containers to keepalive.

With this we can have some benefits:
1) No need to create/delete Pingclient based on the ContainerProxy lifecycle.
2) No need to have multiple actors(even if the overheads for multiple actors are small).

But what I suggested is a small enhancement point regarding performance and it does not have a big difference in terms of the functionality. So I am ok with the current version as well.